test/test_linalg.py

```diff
@@ -31,6 +33,125 @@ def fn(a, b):
 
+    def test_norm_dtype(self, device):
+        def run_test_case(input_size, ord, keepdim, from_dtype, to_dtype, compare_dtype):
+            msg = (
+                f'input_size={input_size}, ord={ord}, keepdim={keepdim}, '
+                f'from_dtype={from_dtype}, to_dtype={to_dtype}')
+            input = torch.randn(*input_size, dtype=from_dtype, device=device)
+            result = torch.linalg.norm(input, ord, keepdim=keepdim, dtype=from_dtype)
+            self.assertEqual(result.dtype, from_dtype, msg=msg)
+            result_converted = torch.linalg.norm(input, ord, keepdim=keepdim, dtype=to_dtype)
+            self.assertEqual(result_converted.dtype, to_dtype, msg=msg)
+            self.assertEqual(result.to(compare_dtype), result_converted.to(compare_dtype), msg=msg)
+
+            result_out_converted = torch.empty_like(result_converted)
+            torch.linalg.norm(input, ord, keepdim=keepdim, dtype=to_dtype, out=result_out_converted)
+            self.assertEqual(result_out_converted.dtype, to_dtype, msg=msg)
+            self.assertEqual(result_converted, result_out_converted, msg=msg)
+
+        ord_vector = [0, 1, -1, 2, -2, 3, -3, 4.5, -4.5, inf, -inf, None]
+        ord_matrix = [1, -1, 2, -2, inf, -inf, None]
+        S = 10
+        test_cases = [
+            ((S, ), ord_vector),
+            ((S, S), ord_matrix),
+        ]
+        for keepdim in [True, False]:
+            for input_size, ord_settings in test_cases:
+                for ord in ord_settings:
+                    # float to double
+                    run_test_case(input_size, ord, keepdim, torch.float, torch.double, torch.float)
+                    # double to float
+                    run_test_case(input_size, ord, keepdim, torch.double, torch.double, torch.float)
+
+        # TODO: Once dtype arg is supported in nuclear and frobenius norms, remove the following test
+        #       and  add 'nuc' and 'fro' to ord_matrix above
+        for ord in ['nuc', 'fro']:
+            input = torch.randn(10, 10, device=device)
+            with self.assertRaisesRegex(ValueError, r'dtype argument is currently not supported'):
+                torch.linalg.norm(input, ord, dtype=torch.float)
+
+    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
+    @dtypes(torch.float, torch.double)
+    def test_norm(self, device, dtype):
+        def run_test_case(input, p, dim, keepdim):
+            result = torch.linalg.norm(input, ord, dim, keepdim)
+            input_numpy = input.cpu().numpy()
+            result_numpy = np.linalg.norm(input_numpy, ord, dim, keepdim)
+
+            msg = f'input.size()={input.size()}, ord={ord}, dim={dim}, keepdim={keepdim}, dtype={dtype}'
+            self.assertEqual(result, result_numpy, msg=msg)
+
+            result_out = torch.empty_like(result)
+            torch.linalg.norm(input, ord, dim, keepdim, out=result_out)
+            self.assertEqual(result, result_out, msg=msg)
+
```

Comment:
I think I should add a JIT test as well, similar to what `outer` has, but I'm not sure of the cleanest way to do that. As far as I can tell, I would need to create multiple different scriptable functions, one for each typing combination of `ord` and `dim`. Something like this:
```
def fn_ord_int_dim_int(input, ord: int, dim: int, keepdim: bool):
    ....

def fn_ord_str_dim_int(input, ord: str, dim: int, keepdim: bool):
    ....
...

script_ord_int_dim_int = torch.jit.script(fn_ord_int_dim_int)
script_ord_str_dim_int = torch.jit.script(fn_ord_str_dim_int)
...
```

And then I'd have to check the types of `ord` and `dim` to call the correct script function. Or is there a better way to do this? Would there be any issue with adding `torch.linalg.norm` test cases to `test_jit.py`?

Reply:
We currently test both the JIT and autograd by adding entries here:

https://github.com/pytorch/pytorch/blob/38c7b9a1680c23062ebaac8ac4304635117fdf9c/torch/testing/_internal/common_methods_invocations.py#L217

Reply:
Right, I wasn't sure how I could use `common_methods_invocations.py` since `torch.linalg.norm` is not a Tensor method. But now I see that when I move it to C++, as `linalg_norm`, I'll be able to access that as a Tensor method.

Reply:
@mruberry, when I first moved `norm` to C++, it was available in Python as `torch.linalg_norm`. However, I just noticed the note in `native_functions.yaml` mentioning that `linalg_` functions should be hidden from the user and pointing to the `torch/csrc/api/include/torch/linalg.h` file. So I made the changes required for that, and it works fine.

But now, I'm not sure how to invoke `torch.linalg.norm` from a test in `common_methods_invocations.py`. The first element of each entry is used as a function name, and as far as I can tell, there isn't a simple way to make it call a function from `torch.linalg` instead of `torch`.

Reply:
Makes sense. To unblock this PR you can just put your own autograd and jit tests into test_linalg.py. While that happens I'll work on a PR that will let you do this automatically.